### DISCUSSION TITLE
Send and receive raw bytes to a raw printer

### DISCUSSION THREAD

**Discussion by wolflu05** [OpenPrinting/cups#883] [Q&A]
Hello, I setup a usb printer with cups using the web interface. Printing PDFs like that works fine. I then tried to expose that printer via the jetdirect protocol using xinetd using the following configuration ([source](https://www.freedomit.co.nz/kb-centos/76-using-port-9100-to-print-to-a-printer-queue-on-a-cups-enabled-server)):

```conf
service jetdirect
{
        socket_type = stream
	type = UNLISTED
        protocol = tcp
	port = 9100
        wait = no
        user = root
        server = /bin/lp
        server_args = -d <cups-printer-name> -o raw
        groups = yes
        disable = no
}
```

This already works fine, I can add the printer on my pc and print as if the usb printer is connected to the pc. This also creates jobs for each print request. But is there any way like this to be able to send raw bytes to the printer and receive bytes that the printer responds with? I want to use that to query the printer for the currently inserted paper, ink stats and some custom printer status commands.

**Comment by zdohnal** [comment]
Hi,

AFAIK there is no way how to get more sophisticated data via CUPS socket backend, which connects to printer's port 9100 - even ink supplies have to be done via SNMP, if the device has SNMP support.
Till today I thought the same can be said for protocols using port 9100 as well, but after checking Wiki this seems to be true only for jetdirect protocol - AppSocket protocol seems to be capable to query a printer status (though it is a protocol, not raw bytes, and I'm not sure if there are any public standard defining the protocol).

However, if the printer supports it (either via network or via USB), I would recommend going for IPP (in case of USB via IPP-over-USB protocol) - it does the exactly the same things you would like to get via socket (you simply ask for the printer attributes ). If the printer does not support IPP 2.0+, you can spend your time figuring out ways how to get such queries via socket, or get used to knowing the fact you won't get such data from printer. Or buying a printer which does IPP 2.0+, which are pretty common in last decade.

**Reply by wolflu05** [reply]
Hello @zdohnal , thank you for you quick reply. After reading your comment I'm not really sure what I should do now. I don't want to buy another printer just because the software doesn't work like I want. IPP and all those printing protocols are new to me. But it seems like my printer implements a custom protocol documented in a technical reference guide for the printer. It's not a regular printer, it's a label printer. If I connect the printer to a computer and send `0x1b 0x41` via e.g. python to the printer, it responds with one byte of data containing status information encoded in each bit. 

So to summarize, if the printer is setup using cups it is not possible and will not be possible to send such raw request? Is that correct? Do you by any chance know any alternatives to cups that implement a spool for raw print requests but still are able to process described status requests?

**Reply by michaelrsweet** [reply]
@wolflu05 CUPS does not provide raw access to printers to users/applications - that way lies madness, and is how printing was done in the 1980's. The normal way of handling this is either through a CUPS printer driver (legacy path) or a printer application which knows how to communicate directly with the printer. Remote clients (Windows or otherwise) are expected to send either common file formats (PDF, JPEG, various streaming raster formats, etc.) or printer-ready data such as PCL, ZPL, etc.

There is no way (with CUPS) to provide a bidirectional raw socket to a Windows system.

**Reply by wolflu05** [reply]
Ok, then it seems like cups is not the right choice here. Do you know any software that i can use for a bidirectional communication with a zebra printer (or similar working) to make a usb only printer accessible through a network?

**Reply by michaelrsweet** [reply]
Sorry, no.

**Reply by wolflu05** [reply]
Ok, but thank you anyway, you already helped me a bit by understanding cups.
